### PR TITLE
fix flattening of ID_invalid_pointer

### PR DESCRIPTION
**Pull request by kroening**
This fixes an inconsistency between the SMT2 encoding and the simplifier on one hand, and the bit-level flattening on the other: NULL is not considered invalid.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: e40e87cd37a7f41c7b8a0468f126a46673851b88).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80301713

File: src/analyses/goto_check.cpp [hunk 1]
```diff
@@ -991,7 +991,8 @@ void goto_checkt::pointer_validity_check(
       allocs=disjunction(disjuncts);
     }
 
-    if(flags.is_unknown() || flags.is_null())
+    if(flags.is_unknown() ||
```
Comment (smowton) [inline src/analyses/goto_check.cpp]: Point of curiosity: all of these `is_` tests are `may_be_` tests, right?

**Review by smowton — APPROVED** [review]
Might be worth cleaning up language, otherwise does what it says on the tin (though I'm not familiar enough to quickly know if there are other corresponding changes that would be needed)

File: src/analyses/goto_check.cpp (hunk 1 above)
Comment (kroening) [reply src/analyses/goto_check.cpp]: Yes, indeed.

File: regression/cbmc/Malloc23/test.desc [hunk 2]
```diff
@@ -7,6 +7,6 @@ pointer outside dynamic object bounds in \*p: FAILURE
 pointer outside dynamic object bounds in \*p: FAILURE
 pointer outside dynamic object bounds in p2\[.*1\]: FAILURE
 pointer outside dynamic object bounds in p2\[.*0\]: FAILURE
-\*\* 4 of 36 failed 
+\*\* 4 of .* failed
```
Comment (tautschnig) [inline regression/cbmc/Malloc23/test.desc]: Not a big deal (and obviously similar changes happened elsewhere): could this please be `[0-9]+` instead of `.*`?

File: regression/cbmc/memory_allocation1/test.desc [hunk 3]
```diff
@@ -3,10 +3,10 @@ main.c
 --pointer-check
 ^EXIT=10$
 ^SIGNAL=0$
-\[main.pointer_dereference.[0-9]+\] dereference failure: pointer invalid in \*p: SUCCESS
-\[main.assertion.1\] assertion \*p==42: SUCCESS
-\[main.pointer_dereference.[0-9]+\] dereference failure: pointer invalid in p\[.*1\]: FAILURE
-\[main.assertion.2\] assertion \*\(p\+1\)==42: SUCCESS
+^\[main\.pointer_dereference\.2\] dereference failure: invalid integer address in \*p: SUCCESS$
```
Comment (tautschnig) [inline regression/cbmc/memory_allocation1/test.desc:6]: Huh? I'm finding it hard to reconcile this with the commit message.

File: src/analyses/goto_check.cpp [hunk 4]
```diff
@@ -1082,6 +1077,18 @@ void goto_checkt::pointer_validity_check(
...
+       flags.is_integer_address())
+    {
+      add_guarded_claim(
+        implies_exprt(integer_address(pointer), allocs),
+        "dereference failure: invalid integer address",
+        "pointer dereference",
+        expr.find_source_location(),
+        expr,
+        guard);
+    }
```
Comment (tautschnig) [inline src/analyses/goto_check.cpp:1091]: I suppose this change and what I mourned above just need to be moved into a separate commit.

**Review by tautschnig — COMMENTED** [review]
With the combination of changes I'm finding it hard to see whether the change that the commit message talks about is actually being tested in any way. Please break apart the changes into separate commits.

File: regression/cbmc/Malloc23/test.desc (hunk 2 above)
Comment (kroening) [reply regression/cbmc/Malloc23/test.desc]: done

File: src/analyses/goto_check.cpp (hunk 4 above)
Comment (kroening) [reply src/analyses/goto_check.cpp:1091]: done

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: ea1a2443c2d4b9b3d01e2c7d2b65b27ccc994f37).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80397583
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 060b59c48a570db4a2ff7126851a4463cbffbfef).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80401170
